server/server.go

```diff
@@ -0,0 +1,100 @@
+package server
+
+import (
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+
+	"github.com/hashicorp/vault/api"
+)
+
+// New returns an unstarted HTTP server with health and proxy handlers.
+func New(logger *log.Logger, vaultClient *api.Client) *http.Server {
+	mux := http.ServeMux{}
+	mux.HandleFunc("/_health", healthHandler)
+	mux.HandleFunc("/", proxyHandler(logger, vaultClient))
+	srv := http.Server{
+		Addr:    "127.0.0.1:8200",
+		Handler: &mux,
+	}
+
+	return &srv
+}
+
+// The proxyHandler is based on the Send function from Vault Agent's proxy:
+// https://github.com/hashicorp/vault/blob/22b486b651b8956d32fb24e77cef4050df7094b6/command/agent/cache/api_proxy.go
+func proxyHandler(logger *log.Logger, vaultClient *api.Client) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		switch r.Method {
+		case "GET":
+			cl, err := vaultClient.Clone()
+			if err != nil {
+				http.Error(w, err.Error(), 500)
+				return
+			}
+			cl.SetToken(vaultClient.Token())
+
+			body, err := ioutil.ReadAll(r.Body)
+			if err != nil {
+				http.Error(w, err.Error(), 500)
+				return
+			}
+
+			logger.Printf("Proxying %s %s\n", r.Method, r.URL.Path)
+			resp, err := proxyRequest(cl, r, body)
+			if err != nil {
+				http.Error(w, err.Error(), 502)
+				return
+			}
+
+			err = resp.Write(w)
+			if err != nil {
+				http.Error(w, "failed to write response back to requester", 500)
+				return
+			}
+			logger.Printf("Successfully proxied %s %s\n", r.Method, r.URL.Path)
+		default:
+			http.Error(w, fmt.Sprintf("unsupported method for vault-lambda-extension: %s", r.Method), 400)
+			return
+		}
+	}
+}
+
+func proxyRequest(client *api.Client, r *http.Request, body []byte) (*api.Response, error) {
+	// http.Transport will transparently request gzip and decompress the response, but only if
+	// the client doesn't manually set the header. Removing any Accept-Encoding header allows the
+	// transparent compression to occur.
+	r.Header.Del("Accept-Encoding")
+	client.SetHeaders(r.Header)
+
+	fwReq := client.NewRequest(r.Method, r.URL.Path)
+	fwReq.BodyBytes = body
+
+	query := r.URL.Query()
+	if len(query) > 0 {
+		fwReq.Params = query
+	}
+
+	resp, err := client.RawRequestWithContext(r.Context(), fwReq)
+	if err != nil {
+		errString := fmt.Sprintf("Failed to proxy request: %s", err)
```

Comment:
This is definitely the area that I felt was weakest in the PR, so I welcome any comments on it. Handling errors is a little tricky because we are acting as an HTTP-level proxy, but the error-handling behaviour of the Vault API client is significantly different from that of the underlying Golang HTTP client. I've reworked the error handling in an attempt to improve the behaviour when Vault returns non-OK error codes, but it's quite confusing and error-prone trying to unwind the additional error handling that `api` does.

Reply:
Based on my comment above and some offline discussion, I've now reworked the proxying to be at the HTTP level instead of using the vault client. I think the error handling is a lot more robust and easier to inspect now. LMK what you think.